File: Annual_Macroeconomic_factors.m
%% Housing Group 8 - Annual Macroeconomic Factors (Simple Analysis)
% Goal: Simple, readable analysis per professor's template
% Data: Annual_Macroeconomic_Factors.xlsx (sheet 'in')

clear; clc; close all;

%% 1) Load dataset (readtable)
file = 'Annual_Macroeconomic_Factors.xlsx';
sheet = 'in';
T = readtable(file, 'Sheet', sheet);

% Ensure Date is datetime
if ~isdatetime(T.Date)
    try
        T.Date = datetime(T.Date);
    catch
        T.Date = datetime(T.Date, 'InputFormat','yyyy-MM-dd'); % fallback
    end
end

%% 2) Handle missing/invalid values (simple + safe)
% Coerce numeric columns (ignore Date)
numVars = setdiff(T.Properties.VariableNames, {'Date'});
for i = 1:numel(numVars)
    % If a column is char/cellstr, convert to numeric (str2double)
    if iscell(T.(numVars{i})) || isstring(T.(numVars{i})) || ischar(T.(numVars{i}))
        T.(numVars{i}) = str2double(string(T.(numVars{i})));
    end
end

% Option A (strict, student-simple): drop rows that are entirely missing in numerics
allNumMissing = all(ismissing(T(:, numVars)), 2);
Tclean = T(~allNumMissing, :);

% (Your file has no missing values; these steps are just to match expectations.)

%% 3) Summary statistics (simple, per professor's list) for a few key variables
varsForStats = {'House_Price_Index','Mortgage_Rate','Unemployment_Rate', ...
                'Real_Disposable_Income','Real_GDP'};

fprintf('=== Summary Statistics (ignoring missing) ===\n');
for v = 1:numel(varsForStats)
    x = Tclean.(varsForStats{v});
    x = x(~isnan(x));
    if isempty(x), continue; end
    fprintf('\nVariable: %s\n', varsForStats{v});
    fprintf('  Count: %d\n', numel(x));
    fprintf('  Mean: %.4f\n', mean(x));
    fprintf('  Std:  %.4f\n', std(x));
    fprintf('  Min:  %.4f\n', min(x));
    fprintf('  Med:  %.4f\n', median(x));
    fprintf('  Max:  %.4f\n', max(x));
    fprintf('  Sum:  %.4f\n', sum(x));
end

%% 4) "Mostly sold item / payment method" analogues
% Not applicable for macro data. We'll instead compute simple decadal summaries.

% Add decade column
yr = year(Tclean.Date);
decade = floor(yr/10)*10;
Tclean.Decade = decade;

% Decadal averages for two intuitive metrics
decTbl = groupsummary(Tclean, "Decade", "mean", ["Mortgage_Rate","Unemployment_Rate"]);
decTbl.Properties.VariableNames = {'Decade','GroupCount','Avg_Mortgage_Rate','Avg_Unemployment_Rate'};

%% 5) Visuals (simple and clear)

% Line chart: main macro series over time (separate figures to keep it simple)
figure('Name','House Price Index');
plot(Tclean.Date, Tclean.House_Price_Index, 'LineWidth',1.5); grid on;
xlabel('Year'); ylabel('Index'); title('House Price Index over Time');

figure('Name','Mortgage Rate');
plot(Tclean.Date, Tclean.Mortgage_Rate, 'LineWidth',1.5); grid on;
xlabel('Year'); ylabel('Percent'); title('Mortgage Rate over Time');

figure('Name','Unemployment Rate');
plot(Tclean.Date, Tclean.Unemployment_Rate, 'LineWidth',1.5); grid on;
xlabel('Year'); ylabel('Percent'); title('Unemployment Rate over Time');

figure('Name','Real Disposable Income');
plot(Tclean.Date, Tclean.Real_Disposable_Income, 'LineWidth',1.5); grid on;
xlabel('Year'); ylabel('Dollars (real)'); title('Real Disposable Income over Time');

% Histogram of Mortgage Rate (like "Histogram of Total Spent" in the example)
figure('Name','Histogram - Mortgage Rate');
histogram(Tclean.Mortgage_Rate);
grid on; xlabel('Mortgage Rate (%)'); ylabel('Count');
title('Distribution of Mortgage Rates');

% Bar chart of decadal average Mortgage Rate (analogue to "Bar Chart per Item")
figure('Name','Bar - Avg Mortgage Rate by Decade');
bar(decTbl.Decade, decTbl.Avg_Mortgage_Rate);
grid on; xlabel('Decade'); ylabel('Avg Mortgage Rate (%)');
title('Average Mortgage Rate by Decade');

% Bar chart of decadal average Unemployment Rate
figure('Name','Bar - Avg Unemployment Rate by Decade');
bar(decTbl.Decade, decTbl.Avg_Unemployment_Rate);
grid on; xlabel('Decade'); ylabel('Avg Unemployment Rate (%)');
title('Average Unemployment Rate by Decade');

%% 6) Save cleaned table and decadal summary (easy to drop in Streamlit)
if ~exist('output','dir'), mkdir('output'); end
writetable(Tclean, fullfile('output','Annual_Macro_Clean.csv'));
writetable(decTbl, fullfile('output','Annual_Macro_Decadal_Summary.csv'));

disp('Done. Clean CSV and summaries saved in /output');

File: python_files/Annual_Macroeconomic_Factors.py
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import os
import streamlit as st

# =========================================================
# 1) Function: Load and clean dataset
# =========================================================
def load_data():
    file = "Annual_Macroeconomic_Factors.xlsx"
    sheet = "in"

    # Load excel file
    T = pd.read_excel(file, sheet_name=sheet)

    # Ensure Date is datetime
    if not np.issubdtype(T["Date"].dtype, np.datetime64):
        try:
            T["Date"] = pd.to_datetime(T["Date"])
        except:
            T["Date"] = pd.to_datetime(T["Date"], format="%Y-%m-%d")

    # Clean numeric variables
    numVars = [c for c in T.columns if c != "Date"]
    for col in numVars:
        if T[col].dtype == object:
            T[col] = pd.to_numeric(T[col], errors="coerce")

    # Remove rows where all numeric values are missing
    allNumMissing = T[numVars].isna().all(axis=1)
    Tclean = T[~allNumMissing].copy()

    # Add decade column
    Tclean["Decade"] = (Tclean["Date"].dt.year // 10) * 10

    # Save cleaned data
    os.makedirs("output", exist_ok=True)
    Tclean.to_csv("output/Annual_Macro_Clean.csv", index=False)

    return Tclean


# =========================================================
# 2) Function: Make Streamlit plots
# =========================================================
def make_plots(Tclean):
    st.subheader("📈 Annual Macroeconomic Trends")

    # --- House Price Index ---
    fig, ax = plt.subplots(figsize=(7,4))
    ax.plot(Tclean["Date"], Tclean["House_Price_Index"])
    ax.set_xlabel("Year")
    ax.set_ylabel("Index")
    ax.set_title("House Price Index")
    ax.grid(True)
    st.pyplot(fig)

    # --- Mortgage Rate ---
    fig, ax = plt.subplots(figsize=(7,4))
    ax.plot(Tclean["Date"], Tclean["Mortgage_Rate"])
    ax.set_xlabel("Year")
    ax.set_ylabel("Percent")
    ax.set_title("Mortgage Rate")
    ax.grid(True)
    st.pyplot(fig)

    # --- Unemployment Rate ---
    fig, ax = plt.subplots(figsize=(7,4))
    ax.plot(Tclean["Date"], Tclean["Unemployment_Rate"])
    ax.set_xlabel("Year")
    ax.set_ylabel("Percent")
    ax.set_title("Unemployment Rate")
    ax.grid(True)
    st.pyplot(fig)

    # --- Real Disposable Income ---
    fig, ax = plt.subplots(figsize=(7,4))
    ax.plot(Tclean["Date"], Tclean["Real_Disposable_Income"])
    ax.set_xlabel("Year")
    ax.set_ylabel("Real $")
    ax.set_title("Real Disposable Income")
    ax.grid(True)
    st.pyplot(fig)

    # --- Histogram ---
    fig, ax = plt.subplots(figsize=(7,4))
    ax.hist(Tclean["Mortgage_Rate"], bins=15)
    ax.set_xlabel("Mortgage Rate (%)")
    ax.set_ylabel("Count")
    ax.set_title("Distribution of Mortgage Rates")
    ax.grid(True)
    st.pyplot(fig)

    # --- Decadal Summary ---
    decTbl = (
        Tclean.groupby("Decade")[["Mortgage_Rate", "Unemployment_Rate"]]
        .mean()
        .reset_index()
        .rename(columns={
            "Mortgage_Rate": "Avg_Mortgage_Rate",
            "Unemployment_Rate": "Avg_Unemployment_Rate"
        })
    )

    # Bar charts
    fig, ax = plt.subplots(figsize=(7,4))
    ax.bar(decTbl["Decade"], decTbl["Avg_Mortgage_Rate"])
    ax.set_xlabel("Decade")
    ax.set_ylabel("Avg Mortgage Rate (%)")
    ax.set_title("Avg Mortgage Rate by Decade")
    ax.grid(True)
    st.pyplot(fig)

    fig, ax = plt.subplots(figsize=(7,4))
    ax.bar(decTbl["Decade"], decTbl["Avg_Unemployment_Rate"])
    ax.set_xlabel("Decade")
    ax.set_ylabel("Avg Unemployment (%)")
    ax.set_title("Avg Unemployment Rate by Decade")
    ax.grid(True)
    st.pyplot(fig)

    return decTbl
